data/data/azure/bootstrap/main.tf

```diff
@@ -0,0 +1,156 @@
+
+resource "random_string" "storage_suffix" {
+  length = 5
+  upper = false
+  special = false
+  keepers = {
+        # Generate a new ID only when a new resource group is defined
+        resource_group = "${var.resource_group_name}"
+    }
+}
+
+resource "azurerm_storage_account" "ignition" {
+  name = "ignitiondata${random_string.storage_suffix.result}"
+  resource_group_name      = "${var.resource_group_name}"
+  location                 = "${var.region}"
+  account_tier             = "Standard"
+  account_replication_type = "LRS"
+}
+
+data "azurerm_storage_account_sas" "ignition" {
+  connection_string = "${azurerm_storage_account.ignition.primary_connection_string}"
+  https_only        = true
+
+  resource_types {
+    service   = false
+    container = false
+    object    = true
+  }
+
+  services {
+    blob  = true
+    queue = false
+    table = false
+    file  = false
+  }
+
+  start  = "${substr(timestamp(), 0, 10)}"
+  expiry = "${substr(timeadd(timestamp(), "24h"), 0, 10)}"
+
+  permissions {
+    read    = true
+    list    = true
+    create  = false
+    add     = false
+    delete  = false
+    process = false
+    write   = false
+    update  = false
+  }
+}
+
+resource "azurerm_storage_container" "ignition" {
+  resource_group_name   = "${var.resource_group_name}"
+  name                  = "ignition"
+  storage_account_name  = "${azurerm_storage_account.ignition.name}"
+  container_access_type = "private"
+}
+
+resource "local_file" "ignition_bootstrap" {
```

Comment:
the installer does't persist the module locations between invocations.
ie. create cluster 
resource creation -> sets up terraform in tmp -> runs -> only saves state.
destroy bootstrap -> sets up terraform in tmp -> runs destro -target=module.bootstrap (this file does not exist) and I think that will cause terraform to fail.

Reply:
I haven't gone through destroy yet.
Q: in azure, we create everything in a single resource group. the easiest to clean it up is to just delete the RG, so why going through terraform to destroy the resources?

Reply:
The reason for the local file is that the azure provider does not support uploading from a string, but only from a path. so I have to create the file before.
I haven't tested destroy yet, but can we park this for now, and we will work on the destroy scenario later on?

Reply:
phase 1: installer creates all the assets upfront in go
phase 2: create all the cloud resources
phase 3: wait for control-plane to bootstrap onto the control-plane(master) machines [bootstrap node decides when that has completed]
phase 4: destroy all the cloud resources required for bootstrap node and unlink bootstrap node from lbs.
phase 5: let cluster create itself internally using operators. 

phase 3 requires the targeted destroy.

Reply:
again, since the image is not available on azure, I was unable to get this far.

> destroy bootstrap -> sets up terraform in tmp -> runs destro -target=module.bootstrap 

it's just speculation at this point, which is why I'd like set this aside and we will address is but I would guess that if I create the file in the main, and pass it down to the bootstrap module, terraform wouldn't see the file and not fail the destroy. However, that means we leave all the secrets on disk?
Ideally, we'd not even write them locally and keep them secure in azure keyvault or in a locked down cloud storage, but let's keep that for next iterations...

Reply:
Terraform won't fail when destroying. It will see that the file does not exist hence the actual state matches the desired state so it will update the current state to reflect it and succeed. It should be idempotent, so running destroy over and over that resource should repeatedly succeed
And I don't see anything that'd make it fail on the dependent object https://github.com/terraform-providers/terraform-provider-azurerm/blob/b7fa1b038dc5c6293a0bcf2740de6d65a65b092c/azurerm/resource_arm_storage_blob.go#L639

This seems a limitation of the terraform resource, I'd also be in favour of tackling this in a later iteration may be try to add support for raw content https://github.com/terraform-providers/terraform-provider-azurerm/blob/b7fa1b038dc5c6293a0bcf2740de6d65a65b092c/azurerm/resource_arm_storage_blob.go#L213

Reply:
Resolving this for now. I'll get back to it later on